### ISSUE TITLE
why sched_setaffinity  is called when import torch

### ISSUE THREAD

**Issue by zedong-mt** [pytorch/pytorch#107366]
### why sched_setaffinity  is called when import torch

the example is my_print.py:
import time
import torch

x1 = torch.randn(2048, 4096).to("cuda")
x2 = torch.randn(4096, 2048).to("cuda")
for i in range(0, 10000):
    y = torch.matmul(x1, x2)

### the test case is:
NCCL_IGNORE_CPU_AFFINITY=1 OMP_PROC_BIND=CLOSE  MKL_NUM_THREADS=8  OMP_NUM_THREADS=8 strace python my_print.py > trace.log 2>&1

### the pid is 74804 and  kernel trace result is:
(py38) root@1752dc0ca2df:/workspace# cat trace.log  | grep sched_setaffinity
sched_setaffinity(74804, 8, [0])   = 0
sched_setaffinity(0, 8, [0])            = 0
sched_setaffinity(0, 8, [0])            = 0
sched_setaffinity(0, 8, [0])            = 0

above code will cause cpu 100% and canot use multi cpus, so  cuda kernel  will become slow.

### Versions
torch 2.0.1+cu117

### my conda environment
packages in environment at /opt/conda/envs/py38:

Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
_openmp_mutex             5.1                       1_gnu    defaults
accelerate                0.19.0                   pypi_0    pypi
alembic                   1.11.2                   pypi_0    pypi
anykeystore               0.2                      pypi_0    pypi
apex                      0.9.10.dev0              pypi_0    pypi
blinker                   1.6.2                    pypi_0    pypi
ca-certificates           2023.05.30           h06a4308_0    defaults
certifi                   2023.7.22                pypi_0    pypi
charset-normalizer        3.2.0                    pypi_0    pypi
click                     8.1.6                    pypi_0    pypi
cloudpickle               2.2.1                    pypi_0    pypi
cmake                     3.27.1                   pypi_0    pypi
contourpy                 1.1.0                    pypi_0    pypi
cryptacular               1.6.2                    pypi_0    pypi
cxxfilt                   0.3.0                    pypi_0    pypi
cycler                    0.11.0                   pypi_0    pypi
databricks-cli            0.17.7                   pypi_0    pypi
deepspeed                 0.10.0+unknown           pypi_0    pypi
defusedxml                0.7.1                    pypi_0    pypi
docker                    6.1.3                    pypi_0    pypi
entrypoints               0.4                      pypi_0    pypi
exceptiongroup            1.1.2                    pypi_0    pypi
filelock                  3.12.2                   pypi_0    pypi
flask                     2.3.2                    pypi_0    pypi
fonttools                 4.42.0                   pypi_0    pypi
gitdb                     4.0.10                   pypi_0    pypi
gitpython                 3.1.32                   pypi_0    pypi
greenlet                  2.0.2                    pypi_0    pypi
gunicorn                  20.1.0                   pypi_0    pypi
hjson                     3.1.0                    pypi_0    pypi
hupper                    1.12                     pypi_0    pypi
idna                      3.4                      pypi_0    pypi
importlib-metadata        6.8.0                    pypi_0    pypi
importlib-resources       6.0.1                    pypi_0    pypi
iniconfig                 2.0.0                    pypi_0    pypi
itsdangerous              2.1.2                    pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
joblib                    1.3.2                    pypi_0    pypi
kiwisolver                1.4.4                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1    defaults
libffi                    3.4.4                h6a678d5_0    defaults
libgcc-ng                 11.2.0               h1234567_1    defaults
libgomp                   11.2.0               h1234567_1    defaults
libstdcxx-ng              11.2.0               h1234567_1    defaults
lit                       16.0.6                   pypi_0    pypi
llvm-openmp               14.0.6               h9e868ea_0    defaults
mako                      1.2.4                    pypi_0    pypi
markdown                  3.4.4                    pypi_0    pypi
markupsafe                2.1.3                    pypi_0    pypi
matplotlib                3.7.2                    pypi_0    pypi
mlflow                    2.5.0                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0    defaults
networkx                  3.1                      pypi_0    pypi
ninja                     1.11.1                   pypi_0    pypi
numpy                     1.24.4                   pypi_0    pypi
nvidia-cublas-cu11        11.10.3.66               pypi_0    pypi
nvidia-cuda-cupti-cu11    11.7.101                 pypi_0    pypi
nvidia-cuda-nvrtc-cu11    11.7.99                  pypi_0    pypi
nvidia-cuda-runtime-cu11  11.7.99                  pypi_0    pypi
nvidia-cudnn-cu11         8.5.0.96                 pypi_0    pypi
nvidia-cufft-cu11         10.9.0.58                pypi_0    pypi
nvidia-curand-cu11        10.2.10.91               pypi_0    pypi
nvidia-cusolver-cu11      11.4.0.1                 pypi_0    pypi
nvidia-cusparse-cu11      11.7.4.91                pypi_0    pypi
nvidia-nccl-cu11          2.14.3                   pypi_0    pypi
nvidia-nvtx-cu11          11.7.91                  pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
openssl                   3.0.10               h7f8727e_0    defaults
packaging                 23.1                     pypi_0    pypi
pandas                    2.0.3                    pypi_0    pypi
pastedeploy               3.0.1                    pypi_0    pypi
pbkdf2                    1.3                      pypi_0    pypi
pillow                    10.0.0                   pypi_0    pypi
pip                       23.2.1           py38h06a4308_0    defaults
plaster                   1.1.2                    pypi_0    pypi
plaster-pastedeploy       1.0.1                    pypi_0    pypi
pluggy                    1.2.0                    pypi_0    pypi
prometheus-client         0.17.1                   pypi_0    pypi
protobuf                  4.24.0                   pypi_0    pypi
psutil                    5.9.5                    pypi_0    pypi
py-cpuinfo                9.0.0                    pypi_0    pypi
pyarrow                   12.0.1                   pypi_0    pypi
pybind11                  2.11.1                   pypi_0    pypi
pydantic                  1.10.12                  pypi_0    pypi
pyjwt                     2.8.0                    pypi_0    pypi
pyparsing                 3.0.9                    pypi_0    pypi
pyramid                   2.0.1                    pypi_0    pypi
pyramid-mailer            0.15.1                   pypi_0    pypi
pytest                    7.4.0                    pypi_0    pypi
python                    3.8.17               h955ad1f_0    defaults
python-dateutil           2.8.2                    pypi_0    pypi
python3-openid            3.2.0                    pypi_0    pypi
pytz                      2023.3                   pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
querystring-parser        1.2.4                    pypi_0    pypi
readline                  8.2                  h5eee18b_0    defaults
regex                     2023.8.8                 pypi_0    pypi
repoze-sendmail           4.4.1                    pypi_0    pypi
requests                  2.31.0                   pypi_0    pypi
requests-oauthlib         1.3.1                    pypi_0    pypi
scikit-learn              1.3.0                    pypi_0    pypi
scipy                     1.10.1                   pypi_0    pypi
setuptools                68.0.0           py38h06a4308_0    defaults
six                       1.16.0                   pypi_0    pypi
smmap                     5.0.0                    pypi_0    pypi
sqlalchemy                2.0.19                   pypi_0    pypi
sqlite                    3.41.2               h5eee18b_0    defaults
sqlparse                  0.4.4                    pypi_0    pypi
sympy                     1.12                     pypi_0    pypi
tabulate                  0.9.0                    pypi_0    pypi
threadpoolctl             3.2.0                    pypi_0    pypi
tk                        8.6.12               h1ccaba5_0    defaults
tomli                     2.0.1                    pypi_0    pypi
torch                     2.0.1                    pypi_0    pypi
torchvision               0.15.2                   pypi_0    pypi
tqdm                      4.66.0                   pypi_0    pypi
transaction               3.1.0                    pypi_0    pypi
translationstring         1.4                      pypi_0    pypi
triton                    2.0.0                    pypi_0    pypi
typing-extensions         4.7.1                    pypi_0    pypi
tzdata                    2023.3                   pypi_0    pypi
urllib3                   1.26.16                  pypi_0    pypi
velruse                   1.1.1                    pypi_0    pypi
venusian                  3.0.0                    pypi_0    pypi
webob                     1.8.7                    pypi_0    pypi
websocket-client          1.6.1                    pypi_0    pypi
werkzeug                  2.3.6                    pypi_0    pypi
wheel                     0.38.4           py38h06a4308_0    defaults
wtforms                   3.0.1                    pypi_0    pypi
wtforms-recaptcha         0.3.2                    pypi_0    pypi
xz                        5.4.2                h5eee18b_0    defaults
zipp                      3.16.2                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_0    defaults
zope-deprecation          5.0                      pypi_0    pypi
zope-interface            6.0                      pypi_0    pypi
zope-sqlalchemy           3.0                      pypi_0    pypi

**Comment by ezyang** [comment]
There is some problem with the conda mkl package, see https://github.com/pytorch/pytorch/issues/99625

**Comment by zedong-mt** [comment]
> There is some problem with the conda mkl package, see #99625

#ezyang  
thank you , but I try to solve it with conda install 'llvm-openmp<16',  but it doesn't work for me,  any other suggestions ?

**Comment by zedong-mt** [comment]
> There is some problem with the conda mkl package, see #99625

I think mkl does not  call sched_setaffinity, see as follows:

(py38) root@bd5a409a10b7:/home/mccxadmin/sirnzhang/liam/muDNN# cat /list.log | grep mkl
/opt/conda/envs/py38/lib/libmkl_core.so.2
/opt/conda/envs/py38/lib/libmkl_gnu_thread.so.2
/opt/conda/envs/py38/lib/libmkl_intel_lp64.so.2
(py38) root@bd5a409a10b7:/home/mccxadmin/sirnzhang/liam/muDNN# readelf -a /opt/conda/envs/py38/lib/libmkl_gnu_thread.so.2 | grep sched_setaffinity
(py38) root@bd5a409a10b7:/home/mccxadmin/sirnzhang/liam/muDNN# readelf -a /opt/conda/envs/py38/lib/libmkl_intel_lp64.so.2 | grep sched_setaffinity
(py38) root@bd5a409a10b7:/home/mccxadmin/sirnzhang/liam/muDNN# readelf -a /opt/conda/envs/py38/lib/libmkl_core.so.2 | grep sched_setaffinity

**Comment by zedong-mt** [comment]
> > There is some problem with the conda mkl package, see #99625
> 
> @ezyang thank you , but I try to solve it with conda install 'llvm-openmp<16', but it doesn't work for me, any other suggestions ?

@ezyang   I  confirmed that the cpu affinity is disappeared with this solution,   but my performance problem and shed_setaffinity system call still exists,  maybe other problems, thank you anyway.

**Comment by jgong5** [comment]
> above code will cause cpu 100% and canot use multi cpus, so cuda kernel will become slow

@zedong-mt I'd like to understand your initial problem more. Are you checking GPU kernel performance or CPU? From your description and your test program, it seems to be related to GPU kernels? How is it related to CPU thread settings here?

**Comment by zedong-mt** [comment]
> > above code will cause cpu 100% and canot use multi cpus, so cuda kernel will become slow
> 
> @zedong-mt I'd like to understand your initial problem more. Are you checking GPU kernel performance or CPU? From your description and your test program, it seems to be related to GPU kernels? How is it related to CPU thread settings here?

@jgong5 thanks.  My problem is about end2end performance actually,  as you know, my program(matrix multiplication as above) inlcudes cpu code(copy the data,  put kernel into gpu stream,  wait for the result,   maybe in torch code, cudnn or cuda runtime) and gpu kernel... while using the same matrix multiplication code(cuda api) and torch api,   there is about 50%  latency gap(but I use msight to profile the latency in both cases, they are the same, so I think it is not the gpu kernel performance),   so I found that cpu affinity(all thread run cpu core 0) maybe cause the gap,  cpu bottleneck,  later I fixed the cpu affinity problem with @ezyang's suggestion, all  thread run on all cpu cores(confirmed with taskset  command),  but the problem exists.

**Comment by jgong5** [comment]
> but the problem exists.

Can you elaborate on what problem still exists? Is it still related to the CPU affinity being discussed in this thread or it is the "50% latency" gap caused by other issues?

**Comment by zedong-mt** [comment]
Actually, my perfomance problem is caused by gpu clock changing,   so I  close this problem.
